information/specification.md

```diff
@@ -0,0 +1,706 @@
+banknotes from a wallet that, when added together, have enough value to cover
+the amount required.  Ideally, we'd always be able to select _just enough_ to
+cover the exact amount.  However, given that coins and banknotes have fixed
+values (and cannot be subdivided), it's often _impossible_ to select the exact
+amount required. In such cases, we typically give the recipient _more_ than the
+required amount, and then receive the excess value back as _change_.
+
+Similarly to how a physical wallet holds value in the form of unspent coins and
+banknotes, a Cardano wallet holds value in the form of a set of _unspent
+transaction outputs_ (UTxO), the results of previous transactions that
+transferred money to the wallet. Each output in the UTxO set has a particular
+_value_, and the total value of a wallet is the _sum_ of these values.
+
+When using a Cardano wallet to make a payment, the wallet software must choose
+unspent outputs from the wallet's UTxO set, so that the total value of selected
+outputs is enough to cover the target amount. Just as with physical money,
+outputs from the UTxO cannot be subdivided, and must be spent completely in a
+given transaction.  Consequently, in the case where it's not possible to cover
+the target amount exactly, the wallet software must arrange that change is paid
+back to the wallet.
+
+Coin selection refers to the process of choosing _unspent outputs_ from a
+wallet's UTxO set in order to make one or more payments, and arranging for
+_change_ to be paid back to the wallet by creating one or more _change
+outputs_.
+
+## Why is Coin Selection Non-Trivial?
+
+### Issues
+
+There are a number of issues which make the problem of coin selection more
+complicated than it would initially appear.
+
+ * Each transaction has a _maximum size_, as defined by the protocol. The size
+   of a transaction increases as we add more inputs or outputs.
+
+   Therefore, there's a practical limit on the number of coins we can select
+   for any given transaction.
+
+ * The most obvious strategy for coin selection, which consists of trying to
+   get as close to the requested value as possible, will tend (over time) to
+   create a lot of _dust_: small unspent outputs.
+
+   Dust outputs are a problem, because even if the total value of dust in a
+   wallet is more than enough to cover a given target amount, if we attempt to
+   include that dust in a given transaction, we may run out of space (by
+   reaching the transaction size limit) before we can cover the target amount.
+
+ * The most obvious strategy for paying change, which consists of making a
+   single change output with the exact excess value, will tend (over time) to
+   reduce the size of the UTxO set. This is bad for two reasons:
+
+    1. Having a small UTxO set limits the number of payments that we can make
+       in parallel. Since a single UTxO entry can only be used to pay for a
+       single output, we need at least as many UTxO entries as there are
```

Comment:
> Since a single UTxO entry can only be used to pay for a single output

This is **not** generally true and in particular, not true for Cardano. This is a limitation our coin selection algorithm self-imposed to itself to make the problem easier to solve (so a multi-variable problem becomes multiple independant single-variable problems, easier to solve).

Reply:
> > Since a single UTxO entry can only be used to pay for a single output

@KtorZ wrote:
> This is **not** generally true and in particular, not true for Cardano. This is a limitation our coin selection algorithm self-imposed to itself to make the problem easier to solve (so a multi-variable problem becomes multiple independant single-variable problems, easier to solve).

**Hypothetical question:** is there a good reason that we could not introduce (in some future version of the wallet) a **_third algorithm_** (to fall back on after trying `Random-Improve` and `Largest-First`) that _relaxes_ the above restriction, and just attempts to fulfill the payment (without attempting any other optimizations)?

Reply:
>> Since a single UTxO entry can only be used to pay for a single output

> This is **not** generally true and in particular, not true for Cardano. 

I've made a change which fixes this here: 02711d3cd95a6d5cb446b7239732e4a95d80e2ee.

Reply:
> Hypothetical question: is there a good reason that we could not introduce (in some future version of the wallet) a third algorithm (to fall back on after trying Random-Improve and Largest-First) that relaxes the above restriction, and just attempts to fulfill the payment (without attempting any other optimizations)?

In theory, no, but there has been a fair amount of research and simulations done regarding the current algorithms. Thus introducing a third one isn't necessarily a piece of cake; now, we'd need to see the influence of a third component on the UTxO